src/core/ddsc/src/dds_rhc_default.c

```diff
@@ -2014,418 +1959,326 @@ static bool take_sample_update_conditions (struct dds_rhc_default *rhc, struct t
-                set_sample_info (info_seq + n, inst, sample);
-                values[n] = ddsi_serdata_ref(sample->sample);
-                rhc->n_vsamples--;
-                if (sample->isread)
-                {
-                  inst->nvread--;
-                  rhc->n_vread--;
-                }
-
-                if (--inst->nvsamples > 0)
-                  psample->next = sample1;
-                else
-                  inst->latest = NULL;
-
-                free_sample (rhc, inst, sample);
-
-                if (++n == max_samples)
-                {
-                  break;
-                }
-              }
-              sample = sample1;
-            }
-          }
-
-          if (inst->inv_exists && n < max_samples && (qmask_of_invsample (inst) & qminv) == 0 && (qcmask == 0 || (inst->conds & qcmask) != 0))
-          {
-            struct trigger_info_qcond dummy_trig_qc;
-#ifndef NDEBUG
-            init_trigger_info_qcond (&dummy_trig_qc);
-#endif
-            take_sample_update_conditions (rhc, &pre, &post, &trig_qc, inst, inst->conds, inst->inv_isread);
-            set_sample_info_invsample (info_seq + n, inst);
-            values[n] = ddsi_serdata_ref(inst->tk->m_sample);
-            inst_clear_invsample (rhc, inst, &dummy_trig_qc);
-            ++n;
-          }
-
-          if (n > n_first && inst->isnew)
-          {
-            inst->isnew = 0;
-            rhc->n_new--;
-          }
-
-          if (n > n_first)
-          {
-            /* if nsamples = 0, it won't match anything, so no need to do
-             anything here for drop_instance_noupdate_no_writers */
-            size_t ntriggers = SIZE_MAX;
-            get_trigger_info_cmn (&post.c, inst);
-            update_conditions_locked (rhc, false, &pre, &post, &trig_qc, inst, NULL, &ntriggers);
-          }
-
-          if (inst_is_empty (inst))
-          {
-            remove_inst_from_nonempty_list (rhc, inst);
-
-            if (inst->isdisposed)
-            {
-              rhc->n_not_alive_disposed--;
-            }
-            if (inst->wrcount == 0)
-            {
-              TRACE ("take: iid %"PRIx64" #0,empty,drop\n", iid);
-              if (!inst->isdisposed)
-              {
-                /* disposed has priority over no writers (why not just 2 bits?) */
-                rhc->n_not_alive_no_writers--;
-              }
-              drop_instance_noupdate_no_writers (rhc, inst);
-            }
-          }
-
-          if (n > n_first)
-            patch_generations (info_seq + n_first, n - n_first - 1);
-        }
-        if (iid == handle)
-        {
-          break;
-        }
-      }
+      n += take_w_qminv_inst (rhc, inst, values + n, info_seq + n, max_samples - n, qminv, qcmask, &ntriggers, to_sample, to_invsample);
       inst = inst1;
     }
   }
   TRACE ("take: returning %"PRIu32"\n", n);
   assert (rhc_check_counts_locked (rhc, true, false));
   ddsrt_mutex_unlock (&rhc->lock);
```

Comment:
One would expect that this unlock is using the same conditon (`lock`) as used when taking this lock in the beginning of the function, but for some reason (related to the `lock_samples` function used to get the sample count) that's not the case. Although this PR does not change this (not so intuitive) behaviour, it would be good to add a comment here that explains the reasoning behind this. Same applies to the unlock in `read_w_qinv`.

Reply:
Good point, thanks!